cores/esp8266/core_esp8266_i2s.c

```diff
@@ -30,147 +30,278 @@
-//always a buffer that is being used by the DMA subsystem *right now* and we don't want to be able to write to that
-//simultaneously.
-
-struct slc_queue_item {
-  uint32  blocksize:12;
-  uint32  datalen:12;
-  uint32  unused:5;
-  uint32  sub_sof:1;
-  uint32  eof:1;
-  uint32  owner:1;
-  uint32  buf_ptr;
-  uint32  next_link_ptr;
-};
-
-static uint32_t i2s_slc_queue[SLC_BUF_CNT-1];
-static uint8_t i2s_slc_queue_len;
-static uint32_t *i2s_slc_buf_pntr[SLC_BUF_CNT]; //Pointer to the I2S DMA buffer data
-static struct slc_queue_item i2s_slc_items[SLC_BUF_CNT]; //I2S DMA buffer descriptors
-static uint32_t *i2s_curr_slc_buf=NULL;//current buffer for writing
-static int i2s_curr_slc_buf_pos=0; //position in the current buffer
-static void (*i2s_callback) (void)=0; //Callback function should be defined as 'void ICACHE_FLASH_ATTR function_name()', placing the function in IRAM for faster execution. Avoid long computational tasks in this function, use it to set flags and process later.
-
-bool ICACHE_FLASH_ATTR i2s_is_full(){
-  return (i2s_curr_slc_buf_pos==SLC_BUF_LEN || i2s_curr_slc_buf==NULL) && (i2s_slc_queue_len == 0);
+#define SLC_BUF_CNT (8)  // Number of buffers in the I2S circular buffer
+#define SLC_BUF_LEN (64) // Length of one buffer, in 32-bit words.
+
+// We use a queue to keep track of the DMA buffers that are empty. The ISR
+// will push buffers to the back of the queue, the I2S transmitter will pull
+// them from the front and fill them. For ease, the queue will contain
+// *pointers* to the DMA buffers, not the data itself. The queue depth is
+// one smaller than the amount of buffers we have, because there's always a
+// buffer that is being used by the DMA subsystem *right now* and we don't
+// want to be able to write to that simultaneously.
+
+// For RX, it's a little different.  The buffers in i2s_slc_queue are
+// placed onto the list when they're filled by DMA
+
+typedef struct slc_queue_item {
+  uint32_t                blocksize : 12;
+  uint32_t                datalen   : 12;
+  uint32_t                unused    :  5;
+  uint32_t                sub_sof   :  1;
+  uint32_t                eof       :  1;
+  volatile uint32_t       owner     :  1; // DMA can change this value
+  uint32_t *              buf_ptr;
+  struct slc_queue_item * next_link_ptr;
+  // This is my own way of tracking item ID
+  volatile uint32_t       myid;
+} slc_queue_item_t;
+
+typedef struct i2s_state {
+  uint32_t *       slc_queue[SLC_BUF_CNT];
+  volatile uint8_t slc_queue_len;
+  uint32_t *       slc_buf_pntr[SLC_BUF_CNT]; // Pointer to the I2S DMA buffer data
+  slc_queue_item_t slc_items[SLC_BUF_CNT]; // I2S DMA buffer descriptors
+  uint32_t *       curr_slc_buf; // Current buffer for writing
+  uint32_t         curr_slc_buf_pos; // Position in the current buffer
+  void             (*callback) (void);
+  // Callback function should be defined as 'void ICACHE_FLASH_ATTR function_name()',
```

Comment:
Can the callback really be in Flash? Shouldn't this read "... defined as void ICACHE_RAM_ATTR function_name()"?

Reply:
You may have hit upon the cause of some random I2S crashes I've seen at idle.  The decorator comes from the original author (maybe cut-n-pasted from the mp3 example at some point?).  I just mirrored what was there in the new functions.  I'll go over and adjust and remove as needed before finishing.

Reply:
The actual I2S interrupt handler is defined "ICACHE_**FLASH**ATTR" in master!